Add UTM to Asset download

tomekkowalczyk opened the pull request:
<!-- ## Which branch should I use for my PR?

Assuming that:

a = current major release
b = current minor release
c = future major release

* a.x for any features and enhancements (e.g. 4.x)
* a.b for any bug fixes (e.g. 4.0, 4.1, 4.2)
* c.x for any features, enhancements or bug fixes with backward compatibility breaking changes (e.g. 5.x) -->

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          |  no
| New feature/enhancement? (use the a.x branch)      | yes
| Deprecations?                          | no
| BC breaks? (use the c.x branch)        | no
| Automated tests included?              |
| Related user documentation PR URL      | 
| Related developer documentation PR URL |
| Issue(s) addressed                     | Fixes #... <!-- prefix each issue number with "Fixes #", no need to create an issue if none exists, explain below instead -->

<!--
Additionally (see https://contribute.mautic.org/contributing-to-mautic/developer/code/pull-requests#work-on-your-pull-request):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too.)
 - Features and deprecations must be submitted against the "4.x" branch.
-->

#### Description:
The functionality allows you to register UTM tags from asset links through GET parameters. UTM columns are added to the asset download table, where the data is retrieved.

These parameters can be read from the level of asset download reports.

URL: http://mautic4.local/asset/15:test?utm_source=test2&utm_medium=test3&utm_campaign=test6&utm_id=test1&utm_term=test4&utm_content=test4

![image](https://user-images.githubusercontent.com/39382654/174294721-8a87caf4-ad54-4578-a848-0063450e37a3.png)

Report Asset Downloads
![image](https://user-images.githubusercontent.com/39382654/174294953-35db8e31-9445-4572-8777-3a5b4f8e6cb3.png)







<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/11266"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

tomekkowalczyk committed:
Add UTM to Asset download

escopecz commented on app/bundles/AssetBundle/Entity/Download.php:
```diff
@@ -66,6 +66,32 @@ class Download
      */
     private $email;
 
+    /**
+     * @var string
+     */
+    private $utmCampaign;
+
+    /**
+     * @var string
+     */
+    private $utmContent;
+
+    /**
+     * @var string
+     */
+    private $utmMedium;
+
+    /**
+     * @var string
+     */
+    private $utmSource;
+
+    /**
+     * @var string
+     */
+    private $utmTerm;
```
Let's use typed properties in new code at least. All entity properties can be null if they are not set directly through the constructor.
```suggestion
    private ?string $utmCampaign;

    private ?string $utmContent;

    private ?string $utmMedium;

    private ?string $utmSource;

    private ?string $utmTerm;
```

escopecz commented on app/bundles/AssetBundle/Entity/Download.php:
```diff
@@ -112,6 +138,31 @@ public static function loadMetadata(ORM\ClassMetadata $metadata)
             ->addJoinColumn('email_id', 'id', true, false, 'SET NULL')
             ->build();
+
+        $builder->createField('utmCampaign', 'string')
+            ->columnName('utm_campaign')
+            ->nullable()
+            ->build();
+
+        $builder->createField('utmContent', 'string')
+            ->columnName('utm_content')
+            ->nullable()
+            ->build();
+
+        $builder->createField('utmMedium', 'string')
+            ->columnName('utm_medium')
+            ->nullable()
+            ->build();
+
+        $builder->createField('utmSource', 'string')
+            ->columnName('utm_source')
+            ->nullable()
+            ->build();
+
+        $builder->createField('utmTerm', 'string')
```
Let's use constants over strings wherever possible
```suggestion
        $builder->createField('utmCampaign', Types::STRING)
            ->columnName('utm_campaign')
            ->nullable()
            ->build();

        $builder->createField('utmContent', Types::STRING)
            ->columnName('utm_content')
            ->nullable()
            ->build();

        $builder->createField('utmMedium', Types::STRING)
            ->columnName('utm_medium')
            ->nullable()
            ->build();

        $builder->createField('utmSource', Types::STRING)
            ->columnName('utm_source')
            ->nullable()
            ->build();

        $builder->createField('utmTerm', Types::STRING)
```

escopecz commented on app/bundles/AssetBundle/Model/AssetModel.php:
```diff
@@ -159,6 +159,11 @@ public function trackDownload($asset, $request = null, $code = '200', $systemEnt
 
         $download = new Download();
         $download->setDateDownload(new \Datetime());
+        $download->setUtmCampaign($request->get('utm_campaign') ?? null);
+        $download->setUtmContent($request->get('utm_content') ?? null);
+        $download->setUtmMedium($request->get('utm_medium') ?? null);
+        $download->setUtmSource($request->get('utm_source') ?? null);
+        $download->setUtmTerm($request->get('utm_term') ?? null);
```
Setting default value of `null` like this is not necessary as the get method returns `null` as the default value automatically.
```suggestion
        $download->setUtmCampaign($request->get('utm_campaign'));
        $download->setUtmContent($request->get('utm_content'));
        $download->setUtmMedium($request->get('utm_medium'));
        $download->setUtmSource($request->get('utm_source'));
        $download->setUtmTerm($request->get('utm_term'));
```

escopecz commented on app/bundles/PageBundle/Controller/PublicController.php:
```diff
@@ -500,7 +500,11 @@ public function redirectAction($redirectId)
 
             if (false !== strpos($url, $this->generateUrl('mautic_asset_download'))) {
-                $url .= '?ct='.$ct;
+                if (strpos($url, '&')) {
+                    $url .= '&ct=' . $ct;
+                } else {
+                    $url .= '?ct=' . $ct;
+                }
```
Isn't this if opposite of what it should do? If it contains `&` then append the GET param with `&`? Also, check the warning in the docs for this function: https://www.php.net/manual/en/function.strpos.php

It's better to compare with `false` as the `0` value can cause unwanted behavior.

escopecz commented on app/bundles/PageBundle/Model/TrackableModel.php:
```diff
@@ -363,6 +363,7 @@ protected function extractTrackablesFromHtml($html)
         /** @var \DOMElement $link */
         foreach ($links as $link) {
             $url = $link->getAttribute('href');
+            $url = preg_replace('/{(\S+?)}/i', '', $url);
```
What is the intent behind this replace? Do you expect whitespaces to be in the links? If so, shouldn't those be replaced with `+` or even better use https://www.php.net/manual/en/function.urlencode.php for the values in the GET params?

escopecz commented on app/bundles/ReportBundle/Translations/en_US/messages.ini:
```diff
@@ -23,6 +23,11 @@ mautic.report.field.lead.title="Contact title"
 mautic.report.field.lead.zipcode = "Contact zip code"
 mautic.report.field.modified_by_user="Last modified by"
 mautic.report.field.publish_down="Publish down on"
+mautic.report.field.utm_campaign="UTM Campaign"
+mautic.report.field.utm_content="UTM Content"
+mautic.report.field.utm_medium="UTM Medium"
+mautic.report.field.utm_source="UTM source"
+mautic.report.field.utm_term="UTM Term"
```
Can you use these translation strings from the LeadBundle instead?
```
mautic.lead.report.utm.campaign="UTM campaign"
mautic.lead.report.utm.content="UTM content"
mautic.lead.report.utm.medium="UTM medium"
mautic.lead.report.utm.source="UTM source"
mautic.lead.report.utm.term="UTM term"
```
I understand that this way there is dependency of the ReportBundle on the LeadBundle but this dependency exists already in so many places that the ReportBundle cannot work without LeadBundle anyway. I think we should use what's already there and save some time for our translators.

escopecz reviewed (changes requested):
Thanks for this feature! I made some comments on how to improve it. Please also take a look at failing CI jobs. I did not report the same issues that are covered by PHPSTAN like adding param and return types for the new methods.

RCheesley commented:
@tomekkowalczyk it would help testers a lot if you were able to give a couple of clear steps as to how they might test this PR. I notice you already have some feedback from code review to address as well :)

RCheesley committed:
Merge branch '4.3' into utm-asset-download

RCheesley commented:
In testing this PR, I was not getting any result at all in the asset downloads table despite the asset showing four downloads:
![mautic ddev site_8037 _ db _ db _ assets _ phpMyAdmin 5 1 3](https://user-images.githubusercontent.com/2930593/176215197-db2351ae-7dab-499a-ba06-47c193a3196f.png)
![mautic ddev site_8037 _ db _ db _ asset_downloads _ phpMyAdmin 5 1 3](https://user-images.githubusercontent.com/2930593/176215214-48eee5f8-e982-4a0a-beab-704c84e949f4.png)

I am not sure why this is the case but would appreciate some proper testing instructions before wasting more time trying to understand myself!

tomekkowalczyk commented:
@escopecz, @RCheesley 

I rewrote the solution to version 5.x and also added tests and applied the tips above https://github.com/mautic/mautic/pull/12098
